### DISCUSSION TITLE
installing IDAKLU solver on Windows

### DISCUSSION THREAD

**Discussion by dion-w** [pybamm-team/PyBaMM#1935] [Q&A]
Hello everyone,
i want to install the IDAKLUsolver on my windows 10 system because of issue #1902 with the CasadiSolver. I found a guide in the pybamm documentation although i don't have WSL available and can't actovate it. I guess this is due to me working on a company laptop (although i have admin rights to install software). 

So i tried to install the basics through msys2. I was able to install compilers and start up the convenience script "install_KLU_Sundials.py" but everytime when it comes to the compilation of the libraries it throws a cmake error that it doesn't find "OpenBLAS" (error log see below, detailed version in the attached .log document). So i downloaded and built OpenBLAS and put it in the pybamm directory directly next to the scripts and pybind11 folder but to no avail.
I even put the generated library files of "OpenBLAS" in different places ([like mentioned here for building scipy](https://scipy.github.io/devdocs/building/windows.html#building-openblas)) but as i'm not proficient in these installations, it was more of a trial and error fueled by my internet research.

So does somebody of you has experience with these problem or know how to deal with them?

> /usr/lib/gcc/x86_64-pc-msys/11.2.0/../../../../x86_64-pc-msys/bin/ld: cannot find -lamd
> /usr/lib/gcc/x86_64-pc-msys/11.2.0/../../../../x86_64-pc-msys/bin/ld: cannot find -lcolamd
> /usr/lib/gcc/x86_64-pc-msys/11.2.0/../../../../x86_64-pc-msys/bin/ld: cannot find -lbtf
> /usr/lib/gcc/x86_64-pc-msys/11.2.0/../../../../x86_64-pc-msys/bin/ld: cannot find -lsuitesparseconfig
> collect2: error: ld returned 1 exit status
> make[1]: *** [Makefile:272: /home/dion.w/.local/lib/libklu.1.3.8.dll] Error 1
> make[1]: Leaving directory '/c/Users/dion.w.IAMET/Documents/Codes/PyBaMM/install_KLU_Sundials/SuiteSparse-5.6.0/KLU/Lib'
> make: *** [Makefile:49: install] Error 2
> 100% [........................................................................] 18496945 / 18496945---------- Running CMake prepare ----------------------------------------
> CMake Warning at config/SundialsCMakeMacros.cmake:83 (message):
>   ------------------------------------------------------------------------
> 
>   WARNING: EXAMPLES_ENABLE is deprecated and will be removed in the future.
> 
>   Copying value to EXAMPLES_ENABLE_C.
> 
>   ------------------------------------------------------------------------
> Call Stack (most recent call first):
>   config/SundialsDeprecated.cmake:20 (print_warning)
>   config/SundialsDeprecated.cmake:25 (print_deprecated)
>   CMakeLists.txt:166 (include)
> 
> 
> -- Using int32_t for indices
> -- Determining Fortran name-mangling scheme... OK
> CMake Error at /usr/share/cmake-3.22.1/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
>   Could NOT find BLAS (missing: BLAS_LIBRARIES)
> Call Stack (most recent call first):
>   /usr/share/cmake-3.22.1/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
>   /usr/share/cmake-3.22.1/Modules/FindBLAS.cmake:1337 (find_package_handle_standard_args)
>   /usr/share/cmake-3.22.1/Modules/FindLAPACK.cmake:240 (find_package)
>   /usr/share/cmake-3.22.1/Modules/FindLAPACK.cmake:276 (_lapack_find_dependency)
>   config/SundialsLapack.cmake:21 (find_package)
>   CMakeLists.txt:927 (include)
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/c/Users/dion.w.IAMET/Documents/Codes/PyBaMM/install_KLU_Sundials/build_sundials/CMakeFiles/CMakeOutput.log".
> See also "/c/Users/dion.w.IAMET/Documents/Codes/PyBaMM/install_KLU_Sundials/build_sundials/CMakeFiles/CMakeError.log".
> ---------- Building the sundials ----------------------------------------
> make: *** No rule to make target 'install'.  Stop.

detailed error log:
[error_noBLAS.txt](https://github.com/pybamm-team/PyBaMM/files/8028237/error_noBLAS.txt)

**Comment by martinjrobins** [comment]
hi @dion-w . I'm not familiar with msys2, but the `install_KLU_Sundials.py` just tries to automate the process of downloading and compiling sundials with KLU support. The KLU support in sundials is dependent on suitesparse, which is also downloaded and compiled by the script. 

However, I can see that msys2 has both sundials (https://packages.msys2.org/base/mingw-w64-sundials) and suitesparse (https://packages.msys2.org/package/mingw-w64-x86_64-suitesparse) as possible packages to install, so you might be able to get away with:
a) installing sundials (if the package maintainers have enabled KLU support)
b) installing suitesparse and then manually compiling sundials with KLU support

I'm not sure why you are getting your error about not finding openblas, and it looks like a few other libraries that should be built by the script (amd, colamd) are not being built. My guess is that the CMake files that come bundled in suitesparse do not support your particular environment, and you might have luck looking at the suitesparse documentation (https://github.com/DrTimothyAldenDavis/SuiteSparse)

Hope this helps
